### PR TITLE
Rev gatk public dependency to 4.alpha.2-61-gcdc484c-20160929.184342-1

### PR DESCRIPTION
**Pull request by droazen**
-Updated tools as necessary to adjust to new ReadFilter framework, as
 well as recent refactoring in GATKVariantContextUtils/MathUtils.

-Updated expected outputs for the HaplotypeCaller exact-match tests to
 reflect expected differences in the ReadPosRankSumTest annotations.

-Removed the --keepduplicatereads argument from CalculateTargetCoverage
 in favor of the new engine-level --disableReadFilter argument, and updated
 .wdl scripts appropriately.

### CODE REVIEW

**Comment by droazen** [comment]
@cmnbroad please review changes to read-filter-related code.

@davidbenjamin please review changes related to `GATKVariantContextUtils`/`MathUtils` refactoring

@LeeTL1220 please review wdl changes (I resorted to some bash argument-translation hackery in order to avoid forcing changes to the jsons)

File: scripts/wdl/README.md [hunk 1]
```diff
@@ -45,10 +45,10 @@ Reference used must be the same between PoN and case samples.
 - ``case_gatk_acnv_workflow.target_file`` --  Same target file (NOT bed format) as was used for the PoN creation.  *This must be the same as was used to generate the PoN*.  Please run ``ConvertBedToTargetFile`` to convert a BED file to a target file.
 - ``case_gatk_acnv_workflow.gatk_jar`` --  Absolute path to gatk-protected.jar
 - ``case_gatk_acnv_workflow.PadTargets.mem`` -- (Advanced) Number of GB RAM to use when running PadTargets.  Always use 1, unless your target file is really large
-- ``case_gatk_acnv_workflow.CalculateTargetCoverage.disable_all_read_filters`` -- (Advanced) Disables all read filters when making coverage profile.  This will probably have unexpected results.  Always use "false".
+- ``case_gatk_acnv_workflow.keep_duplicate_reads`` -- (Advanced) Whether coverage calculations should count duplicate reads.  Setting this to "true" has been shown (empirically) to improve performance.  "false" is also supported.
+- ``case_gatk_acnv_workflow.disable_all_read_filters`` -- (Advanced) Disables all read filters when making coverage profile.  This will probably have unexpected results.  Always use "false".
```
Comment (droazen) [inline scripts/wdl/README.md]: Note that the docs here were just out-of-date (didn't reflect actual names in the jsons) -- no actual changes to the jsons are required for these two arguments.

File: scripts/wdl/pon_gatk_workflow.wdl [hunk 2]
```diff
@@ -217,17 +217,20 @@ task CalculateTargetCoverage {
...
                 --groupBy ${grouping} --transform ${transform} --targets ${padded_target_file} --targetInformationColumns FULL \
-                --keepduplicatereads ${keep_duplicate_reads} --input ${input_bam} --reference ${ref_fasta} \
-                --disable_all_read_filters ${disable_all_read_filters} --interval_set_rule UNION --interval_padding 0 \
+                --input ${input_bam} --reference ${ref_fasta} --disableAllReadFilters ${disable_all_read_filters} \
+                $(if [ ${keep_duplicate_reads} = true ]; then echo " --disableReadFilter NOT_DUPLICATE "; else echo ""; fi) \
+                --interval_set_rule UNION --interval_padding 0 \
                 --secondsBetweenProgressUpdates 10.0 --disableSequenceDictionaryValidation ${disable_sequence_dictionary_validation} \
-                --createOutputBamIndex true --help false --version false --verbosity INFO --QUIET false; \
-          else touch ${entity_id}.coverage.tsv; \
+                --createOutputBamIndex true --help false --version false --verbosity INFO --QUIET false
```
Comment (cmnbroad) [inline scripts/wdl/pon_gatk_workflow.wdl]: Unrelated to this PR, but "--createOutputBamIndex true" is redundant (here and in the wdl above) as its the default.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java [hunk 3]
```diff
@@ -291,24 +288,20 @@ private void initializeActiveRegionEvaluationGenotyperEngine() {
...
-                .and(new CountingReadFilter("PRIMARY_ALIGNMENT", ReadFilterLibrary.PRIMARY_ALIGNMENT))
-                .and(new CountingReadFilter("NOT_DUPLICATE", ReadFilterLibrary.NOT_DUPLICATE))
-                .and(new CountingReadFilter("PASSES_VENDOR_QUALITY_CHECK", ReadFilterLibrary.PASSES_VENDOR_QUALITY_CHECK))
-                .and(new CountingReadFilter("GOOD_CIGAR", ReadFilterLibrary.GOOD_CIGAR))
-                .and(new CountingReadFilter("WELLFORMED", new WellformedReadFilter(header)));
+    public static List<ReadFilter> makeStandardHCReadFilters() {
+        List<ReadFilter> filters = new ArrayList<>();
+        filters.add(new MappingQualityReadFilter(READ_QUALITY_FILTER_THRESHOLD));
+        filters.add(ReadFilterLibrary.MAPPING_QUALITY_AVAILABLE);
+        filters.add(ReadFilterLibrary.MAPPED);
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java]: Somewhat unrelated short circuiting comment: it seems like the order of these 3 filters should be reversed.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngineUnitTest.java [hunk 4]
```diff
@@ -44,7 +45,15 @@ public void testIsActive() {
               final ReferenceDataSource ref = new ReferenceFileSource(reference) ) {
 
             final HaplotypeCallerEngine hcEngine = new HaplotypeCallerEngine(hcArgs, reads.getHeader(), reference.getAbsolutePath());
-            final Iterator<GATKRead> readIter = new ReadFilteringIterator(reads.query(paddedShardInterval), HaplotypeCallerEngine.makeStandardHCReadFilter(hcArgs, reads.getHeader()));
+
+            List<ReadFilter> hcFilters = HaplotypeCallerEngine.makeStandardHCReadFilters();
+            hcFilters.forEach(filter -> filter.setHeader(reads.getHeader()));
+            ReadFilter hcCombinedFilter = hcFilters.get(0);
+            for ( int i = 1; i < hcFilters.size(); ++i ) {
+                hcCombinedFilter = hcCombinedFilter.and(hcFilters.get(i));
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngineUnitTest.java]: Any chance you could give `ReadFilter` an `addAll` method, or a constructor that takes in a `List<ReadFilter>`?

**Comment by davidbenjamin** [comment]
@droazen looks good to me.

**Comment by LeeTL1220** [comment]
@droazen Can we just review in person?  I am not sure that what you've implemented preserves original functionality.

**Comment by droazen** [comment]
@LeeTL1220 sure, we can discuss in person

**Comment by droazen** [comment]
Discussed in person with @LeeTL1220, and alleviated his concerns.

File: scripts/wdl/pon_gatk_workflow.wdl (hunk 2 above)
Comment (droazen) [inline scripts/wdl/pon_gatk_workflow.wdl]: True, but in pipeline scripts it's common to give explicit values to arguments, even when they agree with the defaults.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java (hunk 3 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java]: These three filters are very cheap compared to the later filters (like `WellformedReadFilter`), so I think we're good for now in terms of short-circuiting, unless something shows up on the profiler.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngineUnitTest.java (hunk 4 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngineUnitTest.java]: Yes, that's a good idea -- created https://github.com/broadinstitute/gatk/issues/2198 for this task.

**Pull request merged**